subworkflows/local/virusbreakend_calling/main.nf

```diff
@@ -0,0 +1,154 @@
+    ch_inputs_sorted = ch_tumor_bam
+        .map { meta, tumor_bam, tumor_bai ->
+            return [
+                meta,
+                Utils.selectCurrentOrExisting(tumor_bam, meta, Constants.INPUT.BAM_MARKDUPS_DNA_TUMOR),
+                Utils.selectCurrentOrExisting(tumor_bai, meta, Constants.INPUT.BAI_DNA_TUMOR),
+            ]
+        }
+        .branch { meta, tumor_bam, tumor_bai ->
+            def has_existing = Utils.hasExistingInput(meta, Constants.INPUT.VIRUSINTERPRETER_DIR)
+            runnable: tumor_bam && !has_existing
+            skip: true
+                return meta
+        }
+
+    //
+    // MODULE: VIRUSBreakend
+    //
+    // Create process input channel
+    // channel: [ meta_virus, tumor_bam ]
+    ch_virusbreakend_inputs = ch_inputs_sorted.runnable
+        .map { meta, tumor_bam, tumor_bai ->
+
+            def meta_virus = [
+                key: meta.group_id,
+                id: meta.group_id,
+                sample_id: Utils.getTumorDnaSampleName(meta),
+            ]
+
+            return [meta_virus, tumor_bam]
+        }
+
+    // Run process
+    VIRUSBREAKEND(
+        ch_virusbreakend_inputs,
+        genome_fasta,
+        genome_fai,
+        genome_dict,
+        genome_gridss_index,
+        virusbreakenddb,
+        gridss_config,
+    )
+
+    ch_versions = ch_versions.mix(VIRUSBREAKEND.out.versions)
+
+    //
+    // MODULE: Virus Interpreter
+    //
+    // Select input sources
+    // channel: [ meta, virus_tsv, purple_dir, metrics ]
+    ch_virusinterpreter_inputs_selected = WorkflowOncoanalyser.groupByMeta(
+        WorkflowOncoanalyser.restoreMeta(VIRUSBREAKEND.out.tsv, ch_inputs),
+        ch_purple,
+        ch_bamtools_somatic,
+    )
+        .map { meta, virus_tsv, purple_dir, metrics ->
+
+            def inputs = [
+                virus_tsv,
+                Utils.selectCurrentOrExisting(purple_dir, meta, Constants.INPUT.PURPLE_DIR),
+                Utils.selectCurrentOrExisting(metrics, meta, Constants.INPUT.BAMTOOLS_TUMOR),
+            ]
+
+            return [meta, *inputs]
+        }
+
+    // Sort inputs
+    // channel: [ meta, virus_tsv, purple_dir, metrics ]
+    // channel: skip: [ meta ]
+    ch_virusinterpreter_inputs_sorted = ch_virusinterpreter_inputs_selected
+        .branch { meta, virus_tsv, purple_dir, metrics ->
+            runnable: virus_tsv && purple_dir && metrics
+            skip: true
+                return meta
+        }
+
+    // Create process input channel
+    // channel: [ meta_virus, virus_tsv, purple_dir, metrics ]
+    ch_virusinterpreter_inputs = ch_virusinterpreter_inputs_sorted.runnable
+        .map { d ->
+
+            def meta = d[0]
+            def inputs = d[1..-1]
+
+            def meta_virus = [
+                key: meta.group_id,
+                id: meta.group_id,
+                sample_id: Utils.getTumorDnaSampleName(meta),
+            ]
+
+            return [meta_virus, *inputs]
```

Comment:
Could we possibly remove all the uses of `*` to spread the arguements like this? It's clever, but a couple of us had to look up what it was doing, and we try to keep the groovy as simple as possible so that others can contribute.



https://stackoverflow.com/questions/8790403/asterisks-in-front-of-array-names-in-groovy

This stackoverflow says it was supposed to be removed in Groovy 2.0, but I think Nextflow just moved to 4.0...

Reply:
Hi @edmundmiller, thanks for the feedback!

I'd prefer to keep the spread operator since I find it helps with readability as the resulting type is explicitly shown and the intention is more clear imo e.g. `[*a, *b]` v. `a + b`.

I also checked with the [latest Groovy documentation](https://groovy-lang.org/operators.html#_spread_operator) and the spread operator is still well described. And to confirm that it hasn't been deprecated/removed I compiled a test example use of the spread operator with Groovy 4.0.22 and 5.0.0-alpha-9:

```bash
wget https://groovy.jfrog.io/artifactory/dist-release-local/groovy-zips/apache-groovy-binary-4.0.22.zip && unzip apache-groovy-binary-4.0.22.zip &
wget https://groovy.jfrog.io/artifactory/dist-release-local/groovy-zips/apache-groovy-binary-5.0.0-alpha-9.zip && unzip apache-groovy-binary-5.0.0-alpha-9.zip &
wait

cat <<EOF > spread-operator.groovy
def a = [1, 2, 3]
def b = [4, 5, 6]
def c = [*a, *b]

// NOTE(SW): expected: [1, 2, 3, 4, 5, 6]
println c
EOF
```

```text
$ ./groovy-4.0.22/bin/groovy spread-operator.groovy
[1, 2, 3, 4, 5, 6]

$ ./groovy-5.0.0-alpha-9/bin/groovy spread-operator.groovy
[1, 2, 3, 4, 5, 6]
```

I do take your point where a NF dev/contributor is unaware of the spread operator and is unable or unmotivated to read the relevant Groovy documentation for whatever reason. Though I would think that if a dev/contributor can grok the [NF operators](https://www.nextflow.io/docs/latest/operator.html) then they would find the spread operator trivial in comparison, as you've found yourself.

Ultimately I'd rather promote awareness of useful language features/syntax/concepts rather than avoiding them, though if there is a consensus view amongst the core team that certain parts of the Groovy language should be disallowed or banned across nf-core pipelines then I'll definitely conform to that request.

In the case there isn't yet a consensus view on this, would you find it acceptable to keep use of the spread operator here as is?

Reply:
I would suggest adding part of this explanation into the developer notes that you made to explain the subworkflow structure

Reply:
That's a good idea if we keep this in

Reply:
I haven't yet decided on the most appropriate location for these notes. Tossing up between

- a section in docs/usage.md, or
- expendable section in README.md

Reply:
Any thoughts?

Reply:
Created an issue to continue discussion or revisit at a later time:

- https://github.com/nf-core/oncoanalyser/issues/82

Reply:
And another issue re documentation:

- https://github.com/nf-core/oncoanalyser/issues/84

Reply:
I'm going to mark this resolved, let's move any further discussion to the relevant GH issue!